File: frontend/packages/openmates-cli/src/projectFileRisk.ts
/*
 * Project file-risk classifier for CLI remote-access safety checks.
 *
 * Purpose: keep high-risk write/read exclusions deterministic before any LLM
 * reasoning or future patch-apply flow is considered.
 * Architecture: mirrors backend/shared/python_utils/project_file_risk.py.
 * Security: built-in and user-protected sensitive paths require explicit user
 * approval; this module never silently weakens Project policy.
 * Tests: frontend/packages/openmates-cli/tests/remoteAccess.test.ts.
 */

export interface ProjectFileRiskResult {
  isHighRisk: boolean;
  reasons: string[];
}

const SECRET_OR_ENVIRONMENT_PATTERNS = [
  ".env",
  ".env.*",
  "**/.env",
  "**/.env.*",
];

const DEPLOYMENT_OR_CONFIG_PATTERNS = [
  "Caddyfile",
  "Dockerfile",
  "docker-compose.yml",
  "docker-compose.*.yml",
  "**/terraform/**",
  "**/*.tf",
];

const DEPENDENCY_EXECUTION_PATTERNS = [
  "package.json",
  "package-lock.json",
  "pnpm-lock.yaml",
  "yarn.lock",
  "requirements.txt",
  "pyproject.toml",
  "poetry.lock",
];

const AUTH_SECURITY_PRIVACY_PATTERNS = [
  "**/auth/**",
  "**/security/**",
  "**/privacy/**",
  "**/compliance/**",
  "**/*auth*",
  "**/*security*",
  "**/*privacy*",
];

const MIGRATION_SCHEMA_CI_PATTERNS = [
  "**/migrations/**",
  "**/schemas/**",
  ".github/workflows/**",
  ".gitignore",
];

const BUILT_IN_PATTERNS: Array<{ reason: string; patterns: string[] }> = [
  { reason: "secret_or_environment_file", patterns: SECRET_OR_ENVIRONMENT_PATTERNS },
  { reason: "deployment_or_config_file", patterns: DEPLOYMENT_OR_CONFIG_PATTERNS },
  { reason: "dependency_or_execution_file", patterns: DEPENDENCY_EXECUTION_PATTERNS },
  { reason: "auth_security_privacy_or_compliance_file", patterns: AUTH_SECURITY_PRIVACY_PATTERNS },
  { reason: "migration_schema_or_ci_file", patterns: MIGRATION_SCHEMA_CI_PATTERNS },
];

export function classifyProjectFileRisk(path: string, userProtectedPatterns: string[] = []): ProjectFileRiskResult {
  const normalizedPath = normalizePath(path);
  const reasons: string[] = [];

  for (const group of BUILT_IN_PATTERNS) {
    if (group.patterns.some((pattern) => matchesPattern(normalizedPath, pattern))) {
      reasons.push(group.reason);
    }
  }

  if (userProtectedPatterns.some((pattern) => matchesPattern(normalizedPath, normalizePath(pattern)))) {
    reasons.push("user_protected_pattern");
  }

  return { isHighRisk: reasons.length > 0, reasons: [...new Set(reasons)] };
}

function normalizePath(path: string): string {
  return path.replace(/\\/g, "/").replace(/^\.\//, "");
}

function matchesPattern(path: string, pattern: string): boolean {
  const regex = new RegExp(`^${globToRegex(pattern)}$`);
  return regex.test(path);
}

function globToRegex(pattern: string): string {
  let output = "";
  for (let index = 0; index < pattern.length; index += 1) {
    const char = pattern[index];
    const next = pattern[index + 1];
    if (char === "*" && next === "*") {
      output += ".*";
      index += 1;
    } else if (char === "*") {
      output += "[^/]*";
    } else {
      output += escapeRegex(char ?? "");
    }
  }
  return output;
}

function escapeRegex(value: string): string {
  return value.replace(/[|\\{}()[\]^$+?.]/g, "\\$&");
}

File: backend/shared/python_utils/project_file_risk.py
"""Deterministic file-risk classification for Project remote writes.

Remote Project edits use this classifier before any later-phase write can be
auto-approved. The list intentionally stays narrow: it protects secrets,
deployment, dependency execution, auth/security/privacy/compliance, schema, CI,
and filesystem-boundary risks without turning normal source edits into noise.
"""

from __future__ import annotations

from dataclasses import dataclass
from fnmatch import fnmatch
from pathlib import PurePosixPath


HIGH_RISK_PATTERNS = (
    ".env",
    ".env.*",
    "*.pem",
    "*.key",
    "id_rsa",
    "id_ed25519",
    ".npmrc",
    ".pypirc",
    ".netrc",
    ".ssh/**",
    ".gitignore",
    ".gitattributes",
    ".gitmodules",
    ".github/workflows/**",
    "Caddyfile",
    "nginx.conf",
    "Dockerfile",
    "docker-compose*.yml",
    "k8s/**",
    "helm/**",
    "terraform/**",
    "*.tf",
    "package.json",
    "package-lock.json",
    "pnpm-lock.yaml",
    "yarn.lock",
    "bun.lockb",
    "pnpm-workspace.yaml",
    "pyproject.toml",
    "Cargo.toml",
    "go.mod",
    "migrations/**",
    "schema.sql",
    "backend/core/directus/schemas/**",
)

HIGH_RISK_KEYWORDS = (
    "auth",
    "billing",
    "compliance",
    "encryption",
    "permission",
    "privacy",
    "security",
)


@dataclass(frozen=True)
class ProjectFileRisk:
    is_high_risk: bool
    reasons: list[str]


def _normalize_path(path: str) -> str:
    normalized = path.replace("\\", "/").strip()
    while normalized.startswith("./"):
        normalized = normalized[2:]
    return str(PurePosixPath(normalized)) if normalized else ""


def _matches_pattern(path: str, pattern: str) -> bool:
    basename = PurePosixPath(path).name
    return fnmatch(path, pattern) or fnmatch(basename, pattern)


def classify_project_file_risk(
    path: str,
    *,
    user_protected_patterns: list[str] | None = None,
    is_gitignored: bool = False,
    is_binary: bool = False,
    operation: str = "edit",
) -> ProjectFileRisk:
    """Return whether a Project file operation should always ask by default."""
    normalized_path = _normalize_path(path)
    reasons: list[str] = []

    if not normalized_path or normalized_path.startswith("../"):
        reasons.append("outside_source_root")

    for pattern in HIGH_RISK_PATTERNS:
        if _matches_pattern(normalized_path, pattern):
            reasons.append("built_in_pattern")
            break

    path_parts = {part.lower() for part in PurePosixPath(normalized_path).parts}
    if any(keyword in path_parts for keyword in HIGH_RISK_KEYWORDS):
        reasons.append("sensitive_path_keyword")

    for pattern in user_protected_patterns or []:
        if _matches_pattern(normalized_path, pattern):
            reasons.append("user_protected_pattern")
            break

    if is_gitignored:
        reasons.append("gitignored")
    if is_binary:
        reasons.append("binary")
    if operation in {"delete", "rename", "chmod", "binary_edit", "symlink_edit"}:
        reasons.append(operation)

    return ProjectFileRisk(is_high_risk=bool(reasons), reasons=reasons)
